src/accelerate/data_loader.py

```diff
@@ -64,6 +64,40 @@
         _PYTORCH_DATALOADER_KWARGS.update(additional_kwargs)
 
 
+class SeedableRandomSampler(RandomSampler):
+    """
+    Same as a random sampler, except that in `__iter__` a seed can be used.
+
+    Needed specifically in distributed cases, when the random generator for each GPU needs to start from the same seed
+    and be fully reproducable on multiple iterations.
+
+    If a custom `generator` is passed, it will rely on its initial seed as well as the current iteration it is on
+    (stored in `self.epoch`).
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.epoch = 0
+
+    def __iter__(self):
+        g = torch.Generator()
+        if self.generator is not None:
+            seed = self.epoch + self.generator.initial_seed()
+        else:
+            seed = self.epoch
+        g.manual_seed(seed)
+        n = len(self.data_source)
```

Comment:
The implementation starting from here is quite different from the original `RandomSampler.__iter__`. Why is that?

Reply:
Oh I see, I haven't checked different versions. It looks like the change to using `yield from` is already in PyTorch since v1.7.0 though:

https://github.com/pytorch/pytorch/blame/v1.7.0/torch/utils/data/sampler.py

I checked the PR where this was changed:

https://github.com/pytorch/pytorch/pull/40026

Honestly, not sure if we care so much about it. However, I think the changes from this PR would be worthwhile having here as well:

https://github.com/pytorch/pytorch/pull/103339

which would be an argument in favor of copying the current implementation. But I'm fine with either way.

(Too bad that PyTorch has not moved the seeding to a separate method, which would allow us to override it and leave `__iter__` as is. With the given implementation, I don't think we can seed and call `super().__iter__()`.)

Reply:
Fair point, brought it back in. We sadly can't just call `super().__iter__()`, so copy/pasted with a reference of what is going on/where

Reply:
Note that this still does not contain the change from https://github.com/pytorch/pytorch/pull/103339, which uses `map(int, ...)` instead of `tolist()`. This is to avoid eagerly creating a Python list, which is very inefficient for big `n`.